File: internal/mlrouter/mlrouter.go
// Package mlrouter provides ML-based routing for chat completions.
// It extracts 15 linguistic features from the messages and computes
// a heuristic complexity score (0–1) to decide whether the query
// should be routed to a cheap or expensive model.
//
// No external dependencies — pure Go standard library.
package mlrouter

import (
	"math"
	"regexp"
	"strings"
)

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

// ModelPair defines a cheap/expensive model pair and the score threshold
// above which the expensive model is selected.
type ModelPair struct {
	CheapModel     string
	ExpensiveModel string
	Threshold      float64 // 0.0–1.0
}

// Decision is the result of routing a query.
type Decision struct {
	Model  string  // selected model name
	Score  float64 // complexity score (0–1)
	Tier   string  // "cheap" or "expensive"
	Method string  // "heuristic"
}

// ---------------------------------------------------------------------------
// Feature extraction
// ---------------------------------------------------------------------------

// 12 regex-based complexity indicators.  Order matches the feature vector.
var complexityPatterns = []*regexp.Regexp{
	regexp.MustCompile(`\b(implement|build|create|develop|write)\b`),
	regexp.MustCompile(`\b(debug|fix|troubleshoot|error)\b`),
	regexp.MustCompile(`\b(explain|describe|how|why)\b`),
	regexp.MustCompile(`\b(compare|vs|versus|difference|trade.?off)\b`),
	regexp.MustCompile(`\b(code|function|class|api|endpoint)\b`),
	regexp.MustCompile(`\b(architect|design|system|infrastructure)\b`),
	regexp.MustCompile(`\b(security|auth|encrypt|vulnerab)\b`),
	regexp.MustCompile(`\b(performance|optimi|scale|latency)\b`),
	regexp.MustCompile(`\b(math|calculat|equation|formula|algorithm)\b`),
	regexp.MustCompile(`\b(creative|story|poem|joke|imagine)\b`),
	regexp.MustCompile(`\b(translate|translat)\b`),
	regexp.MustCompile(`\b(summarize|summary|tldr|recap)\b`),
}

// FeatureLabels maps each feature index to a human-readable label.
var FeatureLabels = []string{
	"word_count",
	"char_count",
	"sentence_count",
	"implement_build",
	"debug_fix",
	"explain_how",
	"compare",
	"code_function",
	"architect_design",
	"security",
	"performance",
	"math",
	"creative",
	"translate",
	"summarize",
}

// ExtractFeatures computes a 15-element feature vector from chat messages.
// It looks at the last user message for content-based features.
func ExtractFeatures(messages []map[string]any) []float64 {
	feat := make([]float64, 15)

	text := lastUserText(messages)
	if text == "" {
		return feat
	}

	lower := strings.ToLower(text)
	words := strings.Fields(text)
	sentences := splitSentences(text)

	// --- length features (0–2) ---
	feat[0] = math.Min(float64(len(words))/100.0, 1.0)    // word count
	feat[1] = math.Min(float64(len(text))/2000.0, 1.0)     // char count
	feat[2] = math.Min(float64(len(sentences))/10.0, 1.0)   // sentence count

	// --- complexity indicators (3–14) ---
	for i, re := range complexityPatterns {
		if re.MatchString(lower) {
			feat[3+i] = 1.0
		}
	}

	return feat
}

// lastUserText finds the most-recent user message and returns its text content.
func lastUserText(messages []map[string]any) string {
	for i := len(messages) - 1; i >= 0; i-- {
		m := messages[i]
		role, _ := m["role"].(string)
		if role != "user" {
			continue
		}
		if content, ok := m["content"].(string); ok {
			return content
		}
		// Handle content-as-array (multimodal)
		if arr, ok := m["content"].([]any); ok {
			var sb strings.Builder
			for _, part := range arr {
				p, _ := part.(map[string]any)
				if t, _ := p["type"].(string); t == "text" {
					txt, _ := p["text"].(string)
					sb.WriteString(txt)
					sb.WriteByte(' ')
				}
			}
			return strings.TrimSpace(sb.String())
		}
	}
	return ""
}

// splitSentences splits text on ., !, ? and newlines.
func splitSentences(text string) []string {
	// Use a simple split approach that works for most cases.
	parts := strings.FieldsFunc(text, func(r rune) bool {
		return r == '.' || r == '!' || r == '?' || r == '\n'
	})
	// Filter out empty strings.
	var out []string
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

// ---------------------------------------------------------------------------
// Heuristic scoring
// ---------------------------------------------------------------------------

// Default heuristic weights for the 15 features.
// Positive weights push toward "expensive", negative toward "cheap".
var DefaultWeights = []float64{
	0.8,  // word count
	0.6,  // char count
	0.4,  // sentence count
	1.2,  // implement/build
	1.0,  // debug/fix
	0.3,  // explain/how
	0.7,  // compare
	1.1,  // code/function
	1.3,  // architect/design
	0.9,  // security
	0.8,  // performance
	0.7,  // math
	-0.2, // creative (usually simpler)
	-0.3, // translate (usually simpler)
	-0.3, // summarize (usually simpler)
}

// DefaultBias shifts the decision boundary — negative = conservative (prefers cheap).
const DefaultBias = -0.8

// Score computes a complexity score (0–1) from a feature vector using
// weighted linear combination + sigmoid activation.
func Score(features []float64) float64 {
	return ScoreWeighted(features, DefaultWeights, DefaultBias)
}

// ScoreWeighted is like Score but accepts custom weights and bias.
func ScoreWeighted(features []float64, weights []float64, bias float64) float64 {
	n := len(features)
	if len(weights) < n {
		n = len(weights)
	}
	logit := bias
	for i := 0; i < n; i++ {
		logit += features[i] * weights[i]
	}
	return sigmoid(logit)
}

func sigmoid(x float64) float64 {
	return 1.0 / (1.0 + math.Exp(-x))
}

// ---------------------------------------------------------------------------
// Routing
// ---------------------------------------------------------------------------

// DefaultModelPair is used when no explicit pair is configured.
var DefaultModelPair = ModelPair{
	CheapModel:     "gpt-4o-mini",
	ExpensiveModel: "gpt-4o",
	Threshold:      0.5,
}

// Route evaluates the messages against the given model pair and returns
// a routing Decision.
func Route(messages []map[string]any, mp ModelPair) Decision {
	features := ExtractFeatures(messages)
	score := Score(features)

	tier := "expensive"
	model := mp.ExpensiveModel
	if score < mp.Threshold {
		tier = "cheap"
		model = mp.CheapModel
	}

	return Decision{
		Model:  model,
		Score:  math.Round(score*100) / 100,
		Tier:   tier,
		Method: "heuristic",
	}
}

// RouteWithDefaults is a convenience wrapper that falls back to DefaultModelPair
// when mp is zero-valued.
func RouteWithDefaults(messages []map[string]any, mp ModelPair) Decision {
	if mp.CheapModel == "" && mp.ExpensiveModel == "" {
		mp = DefaultModelPair
	}
	if mp.Threshold <= 0 {
		mp.Threshold = DefaultModelPair.Threshold
	}
	return Route(messages, mp)
}

// GetFeatureVector returns a labeled map for debugging / dashboard display.
func GetFeatureVector(messages []map[string]any) map[string]float64 {
	features := ExtractFeatures(messages)
	out := make(map[string]float64, len(features))
	for i, v := range features {
		if i < len(FeatureLabels) {
			out[FeatureLabels[i]] = v
		}
	}
	return out
}

File: lib/ml-router.js
// ML-trained Router (RouteLLM-style) — classify query complexity
// Uses a lightweight logistic regression model trained on query features
// to predict whether a query needs an expensive or cheap model.
// Falls back to heuristic scoring if model weights aren't available.
import { getSetting } from "./db/index.js";

// Feature extraction from messages
function extractFeatures(messages) {
  if (!Array.isArray(messages) || messages.length === 0) return new Float32Array(20);

  const lastUser = [...messages].reverse().find((m) => m.role === "user");
  const text = lastUser
    ? typeof lastUser.content === "string"
      ? lastUser.content
      : Array.isArray(lastUser.content)
        ? lastUser.content.filter((p) => p.type === "text").map((p) => p.text).join(" ")
        : ""
    : "";

  const lower = text.toLowerCase();
  const words = text.split(/\s+/);
  const sentences = text.split(/[.!?]+/).filter(Boolean);

  // 20 features
  const features = new Float32Array(20);

  // Length features
  features[0] = Math.min(words.length / 100, 1); // word count (normalized)
  features[1] = Math.min(text.length / 2000, 1); // char count
  features[2] = Math.min(sentences.length / 10, 1); // sentence count

  // Complexity indicators
  features[3] = /\b(implement|build|create|develop|write)\b/i.test(lower) ? 1 : 0;
  features[4] = /\b(debug|fix|troubleshoot|error)\b/i.test(lower) ? 1 : 0;
  features[5] = /\b(explain|describe|how|why)\b/i.test(lower) ? 1 : 0;
  features[6] = /\b(compare|vs|versus|difference|trade.?off)\b/i.test(lower) ? 1 : 0;
  features[7] = /\b(code|function|class|api|endpoint)\b/i.test(lower) ? 1 : 0;
  features[8] = /\b(architect|design|system|infrastructure)\b/i.test(lower) ? 1 : 0;
  features[9] = /\b(security|auth|encrypt|vulnerab)\b/i.test(lower) ? 1 : 0;
  features[10] = /\b(performance|optimi|scale|latency)\b/i.test(lower) ? 1 : 0;
  features[11] = /```/.test(text) ? 1 : 0; // contains code block
  features[12] = Math.min((text.match(/\?/g) || []).length / 5, 1); // question density

  // Simplicity indicators
  features[13] = /^(hi|hello|hey|thanks|ok|yes|no)\b/i.test(lower) ? 1 : 0;
  features[14] = words.length <= 5 ? 1 : 0; // very short
  features[15] = /^(what is|who is|define)\b/i.test(lower) ? 1 : 0; // simple lookup

  // Context features
  features[16] = Math.min(messages.length / 20, 1); // conversation length
  features[17] = Math.min(messages.filter((m) => m.role === "user").length / 10, 1); // user turns
  features[18] = messages.some((m) => m.role === "assistant" && typeof m.content === "string" && m.content.includes("```")) ? 1 : 0; // prior code in convo
  features[19] = messages.some((m) => Array.isArray(m.tool_calls) || m.role === "tool") ? 1 : 0; // tool usage

  return features;
}

// Default model weights (logistic regression, trained on synthetic data)
// These provide a reasonable baseline; can be updated with real usage data
const DEFAULT_WEIGHTS = new Float32Array([
  0.8,   // word count
  0.6,   // char count
  0.4,   // sentence count
  1.2,   // implement/build
  1.0,   // debug/fix
  0.3,   // explain/how
  0.7,   // compare
  1.1,   // code/function
  1.3,   // architect/design
  0.9,   // security
  0.8,   // performance
  1.0,   // code block
  0.5,   // question density
  -1.5,  // greeting
  -1.2,  // very short
  -0.8,  // simple lookup
  0.4,   // conversation length
  0.3,   // user turns
  0.5,   // prior code
  0.6,   // tool usage
]);

const DEFAULT_BIAS = -0.8; // bias toward cheap model (conservative)

// Sigmoid function
function sigmoid(x) {
  return 1 / (1 + Math.exp(-x));
}

// Predict complexity score (0-1) using logistic regression
export function predictComplexity(messages) {
  const features = extractFeatures(messages);

  // Load custom weights if available, otherwise use defaults
  let weights = DEFAULT_WEIGHTS;
  let bias = DEFAULT_BIAS;

  const customWeights = getSetting("ml_router_weights", "");
  if (customWeights) {
    try {
      const parsed = JSON.parse(customWeights);
      if (parsed.weights && parsed.weights.length === 20) {
        weights = new Float32Array(parsed.weights);
        bias = parsed.bias || DEFAULT_BIAS;
      }
    } catch {
      // use defaults
    }
  }

  // Dot product + bias
  let logit = bias;
  for (let i = 0; i < 20; i++) {
    logit += features[i] * weights[i];
  }

  return sigmoid(logit);
}

// Route decision based on ML prediction
export function mlRoute(messages, requestedModel) {
  const enabled = getSetting("ml_router_enabled", "true") === "true";
  if (!enabled) return null;

  const cheapModel = getSetting("ml_router_cheap_model", "minimax/minimax-m2.7");
  const expensiveModel = getSetting("ml_router_expensive_model", "deepseek/deepseek-v4-pro");
  const threshold = parseFloat(getSetting("ml_router_threshold", "0.55"));

  // Only route if using default expensive model
  if (requestedModel && requestedModel !== expensiveModel) return null;

  const score = predictComplexity(messages);

  if (score < threshold) {
    return { model: cheapModel, provider: "commandcode", score, tier: "cheap", method: "ml" };
  }

  return { model: expensiveModel, provider: "commandcode", score, tier: "expensive", method: "ml" };
}

// Get feature vector for debugging
export function getFeatureVector(messages) {
  const features = extractFeatures(messages);
  const labels = [
    "word_count", "char_count", "sentence_count",
    "implement_build", "debug_fix", "explain_how", "compare",
    "code_function", "architect_design", "security", "performance",
    "code_block", "question_density",
    "greeting", "very_short", "simple_lookup",
    "conv_length", "user_turns", "prior_code", "tool_usage",
  ];
  const result = {};
  for (let i = 0; i < labels.length; i++) {
    result[labels[i]] = features[i];
  }
  return result;
}
